src/error.rs

```diff
@@ -52,22 +76,27 @@ impl<T: PartialEq + Display> Display for Located<T> {
     }
 }
 
-#[derive(Debug, PartialEq, Clone)]
-pub struct SchemeError {
-    pub category: ErrorType,
-    pub message: String,
-    pub location: Option<[u32; 2]>,
+#[derive(PartialEq, Error, Clone)]
+pub enum ErrorData {
+    #[error("syntax error: {0}")]
+    Syntax(SyntaxError),
```

Comment:
why not use #[from], then replace all "syntax_error!(<err>)" and "logic_error!(<err>)" with "<err>?" ?
?operator will auto cast Result<_, E1> to return error type Result<_, E2>, if impl From<E1> for E2